### PR TITLE
[extended checker] Add checks for safe randomness usage

### PR DESCRIPTION
**Pull request by wrwg**
## Description

This adds checks to the extended checker for safe usage of randomness.

- Public functions which directly or indirectly call into the `randomness` module are flagged as an error. Exempted functions are those which have the `#[lint_allow_unsafe_randomness]` attribute or which are in the framework at address 0x1.
- Private or friend entry functions which do not have the `#[randomness]` attribute are flagged as an error.

We do not need to check dynamic dispatch here since only public functions are allowed to be registered for dispatch, and those are already checked via above logic.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Added Aptos transactional tests

## Key Areas to Review

Consistency of the checks.

### CODE REVIEW

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/borrow_in_loop.exp [hunk 1]
```diff
@@ -1,7 +1,55 @@
 processed 4 tasks
 
 task 1 'publish'. lines 4-26:
-Error: VMError with status STLOC_UNSAFE_TO_DESTROY_ERROR at location Module ModuleId { address: f75daa73fc071f93593335eb9033da804777eb94491650dd3f095ce6f778acb6, name: Identifier("m") } at index 0 for function definition at code offset 11 in function definition 0
+Error: VMError with status STLOC_UNSAFE_TO_DESTROY_ERROR at location Module ModuleId { address: f75daa73fc071f93593335eb9033da804777eb94491650dd3f095ce6f778acb6, name: Identifier("m") } and message Backtrace: Backtrace [
```
Comment (brmataptos) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/borrow_in_loop.exp:4]: Maybe you accidentally included a debugging stack trace here?

**Review by brmataptos — APPROVED** [review]
I can't comment on the mechanics of the randomness usage check, but rest looks good.

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/borrow_in_loop.exp (hunk 1 above)
Comment (wrwg) [reply aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/borrow_in_loop.exp:4]: Yep, I forgot that I currently always need to do `(unset RUST_BACKTRACE; UB=1 cargo test)`

**Review by vineethk — COMMENTED** [review]
1. What's an appropriate place for documenting the `lint_allow_unsafe_randomness` attribute? Perhaps along with the `randomness` module documentation? The release notes documentation is probably insufficient (although the error message does direct the user what to do to skip the check, so it is not a blocker for this PR).
2. One scenario I wanted to bring up is the following: say we have `A::foo()` which is public and has `#[lint_allow_unsafe_randomness]` and uses randomness features. Now, a public function in a different module, say `B::bar()`, which does not have the unsafe randomness attribute, dynamically dispatches to `A::foo()`. Then we have a scenario where there are no compiler errors anywhere, but we do have a public function calling indirectly into randomness feature. Is this a concern?

Other than these, I have just small nit comments, and the PR LGTM.

File: aptos-move/e2e-move-tests/src/tests/randomness.data/pack/sources/test.move [hunk 2]
```diff
@@ -17,7 +17,8 @@ module 0x1::test {
         let _ = randomness::u64_integer();
     }
 
-    entry fun fail_if_not_annotated_and_using_randomness() {
+    #[lint_allow_unsafe_randomness]
```
Comment (vineethk) [inline aptos-move/e2e-move-tests/src/tests/randomness.data/pack/sources/test.move:20]: I am assuming this is to test that an arbitrary attribute (such as the now unsupported  `#[lint_allow_unsafe_randomness]`) still causes failure?

Or was this supposed to be updated to `#[lint::allow_unsafe_randomness]`?

**Review by vineethk — APPROVED** [review]

File: aptos-move/e2e-move-tests/src/tests/randomness.data/pack/sources/test.move (hunk 2 above)
Comment (wrwg) [reply aptos-move/e2e-move-tests/src/tests/randomness.data/pack/sources/test.move:20]: Oversight, this fails the test. Fixed.

**Pull request merged**